File: packages/core/src/workflow.ts
import { IntentDocument, IntentBlock } from "./types.js";

export interface WorkflowStep {
  block: IntentBlock;
  dependsOn: string[];
  dependedOnBy: string[];
  isGate: boolean;
  isTerminal: boolean;
  isParallel: boolean;
}

export interface WorkflowGraph {
  entryPoints: string[];
  steps: Record<string, WorkflowStep>;
  executionOrder: string[][];
  gatePositions: number[];
  hasTerminal: boolean;
  warnings: string[];
}

const WORKFLOW_TYPES = new Set([
  "step",
  "gate",
  "decision",
  "parallel",
  "wait",
  "result",
  "checkpoint",
  "retry",
  "call",
  "handoff",
  "trigger",
  "loop",
  "signal",
  "error",
  "audit",
]);

function collectWorkflowBlocks(blocks: IntentBlock[]): IntentBlock[] {
  const result: IntentBlock[] = [];
  for (const block of blocks) {
    if (WORKFLOW_TYPES.has(block.type)) {
      result.push(block);
    }
    if (block.children) {
      result.push(...collectWorkflowBlocks(block.children));
    }
  }
  return result;
}

function getStepId(block: IntentBlock): string {
  return (block.properties?.id as string) || block.id;
}

function isInsideParallel(block: IntentBlock, doc: IntentDocument): boolean {
  function search(blocks: IntentBlock[], insideParallel: boolean): boolean {
    for (const b of blocks) {
      const inP = insideParallel || b.type === "parallel";
      if (b === block) return inP;
      if (b.children) {
        const found = search(b.children, inP);
        if (found) return true;
      }
    }
    return false;
  }
  return search(doc.blocks, false);
}

export function extractWorkflow(doc: IntentDocument): WorkflowGraph {
  if (!doc || !doc.blocks) {
    return {
      entryPoints: [],
      steps: {},
      executionOrder: [],
      gatePositions: [],
      hasTerminal: false,
      warnings: ["Empty or invalid document"],
    };
  }

  const warnings: string[] = [];
  const workflowBlocks = collectWorkflowBlocks(doc.blocks);

  if (workflowBlocks.length === 0) {
    return {
      entryPoints: [],
      steps: {},
      executionOrder: [],
      gatePositions: [],
      hasTerminal: false,
      warnings: ["No workflow blocks found"],
    };
  }

  // Build step map
  const steps: Record<string, WorkflowStep> = {};
  const idMap = new Map<string, IntentBlock>(); // track explicit IDs

  for (const block of workflowBlocks) {
    const stepId = getStepId(block);

    if (steps[stepId]) {
      warnings.push(`Duplicate step ID: ${stepId}`);
      continue;
    }

    idMap.set(stepId, block);
    steps[stepId] = {
      block,
      dependsOn: [],
      dependedOnBy: [],
      isGate: block.type === "gate",
      isTerminal: block.type === "result",
      isParallel: isInsideParallel(block, doc),
    };
  }

  // Build dependency graph from depends: property
  for (const [stepId, step] of Object.entries(steps)) {
    const depends = step.block.properties?.depends;
    if (depends && typeof depends === "string") {
      const deps = depends
        .split(",")
        .map((d) => d.trim())
        .filter(Boolean);
      for (const dep of deps) {
        if (!steps[dep]) {
          warnings.push(`Step "${stepId}" depends on unknown step "${dep}"`);
          continue;
        }
        step.dependsOn.push(dep);
        steps[dep].dependedOnBy.push(stepId);
      }
    }
  }

  // Topological sort (Kahn's algorithm)
  const inDegree: Record<string, number> = {};
  for (const id of Object.keys(steps)) {
    inDegree[id] = steps[id].dependsOn.length;
  }

  const executionOrder: string[][] = [];
  const processed = new Set<string>();

  let iterationGuard = 0;
  const maxIterations = Object.keys(steps).length + 1;

  while (
    processed.size < Object.keys(steps).length &&
    iterationGuard < maxIterations
  ) {
    iterationGuard++;

    // Collect all steps with in-degree 0 that haven't been processed
    const batch: string[] = [];
    for (const id of Object.keys(steps)) {
      if (!processed.has(id) && inDegree[id] === 0) {
        batch.push(id);
      }
    }

    if (batch.length === 0) {
      // Cycle detected - add remaining steps with a warning
      const remaining = Object.keys(steps).filter((id) => !processed.has(id));
      warnings.push(`Cycle detected involving: ${remaining.join(", ")}`);
      executionOrder.push(remaining);
      for (const id of remaining) {
        processed.add(id);
      }
      break;
    }

    executionOrder.push(batch);
    for (const id of batch) {
      processed.add(id);
      // Reduce in-degree for dependents
      for (const dependent of steps[id].dependedOnBy) {
        inDegree[dependent]--;
      }
    }
  }

  // Entry points = steps with no dependencies
  const entryPoints = Object.keys(steps).filter(
    (id) => steps[id].dependsOn.length === 0,
  );

  // Gate positions = batch indices that contain a gate
  const gatePositions: number[] = [];
  for (let i = 0; i < executionOrder.length; i++) {
    if (executionOrder[i].some((id) => steps[id]?.isGate)) {
      gatePositions.push(i);
    }
  }

  const hasTerminal = Object.values(steps).some((s) => s.isTerminal);

  return {
    entryPoints,
    steps,
    executionOrder,
    gatePositions,
    hasTerminal,
    warnings,
  };
}

File: packages/rust/src/workflow.rs
//! Workflow graph extraction.
//!
//! Parity target: packages/core/src/workflow.ts

use std::collections::{HashMap, HashSet};

use crate::types::{IntentBlock, IntentDocument};

#[derive(Debug, Clone)]
pub struct WorkflowStep {
    pub block: IntentBlock,
    pub depends_on: Vec<String>,
    pub depended_on_by: Vec<String>,
    pub is_gate: bool,
    pub is_terminal: bool,
    pub is_parallel: bool,
}

#[derive(Debug, Clone)]
pub struct WorkflowGraph {
    pub entry_points: Vec<String>,
    pub steps: HashMap<String, WorkflowStep>,
    pub execution_order: Vec<Vec<String>>,
    pub gate_positions: Vec<usize>,
    pub has_terminal: bool,
    pub warnings: Vec<String>,
}

const WORKFLOW_TYPES: &[&str] = &[
    "step",
    "gate",
    "decision",
    "parallel",
    "wait",
    "result",
    "checkpoint",
    "retry",
    "call",
    "handoff",
    "trigger",
    "loop",
    "signal",
    "error",
    "audit",
];

pub fn extract_workflow(doc: &IntentDocument) -> WorkflowGraph {
    if doc.blocks.is_empty() {
        return WorkflowGraph {
            entry_points: vec![],
            steps: HashMap::new(),
            execution_order: vec![],
            gate_positions: vec![],
            has_terminal: false,
            warnings: vec!["Empty or invalid document".to_string()],
        };
    }

    let mut warnings = Vec::new();
    let mut workflow_blocks: Vec<(IntentBlock, bool)> = Vec::new();
    collect_workflow_blocks(&doc.blocks, false, &mut workflow_blocks);

    if workflow_blocks.is_empty() {
        return WorkflowGraph {
            entry_points: vec![],
            steps: HashMap::new(),
            execution_order: vec![],
            gate_positions: vec![],
            has_terminal: false,
            warnings: vec!["No workflow blocks found".to_string()],
        };
    }

    let mut steps: HashMap<String, WorkflowStep> = HashMap::new();
    for (block, is_parallel) in workflow_blocks {
        let step_id = block
            .properties
            .as_ref()
            .and_then(|p| p.get("id").cloned())
            .unwrap_or_else(|| block.id.clone());

        if steps.contains_key(&step_id) {
            warnings.push(format!("Duplicate step ID: {step_id}"));
            continue;
        }

        steps.insert(
            step_id,
            WorkflowStep {
                is_gate: block.block_type == "gate",
                is_terminal: block.block_type == "result",
                block,
                depends_on: vec![],
                depended_on_by: vec![],
                is_parallel,
            },
        );
    }

    let ids: Vec<String> = steps.keys().cloned().collect();
    for id in &ids {
        let depends_raw = steps
            .get(id)
            .and_then(|s| s.block.properties.as_ref())
            .and_then(|p| p.get("depends").cloned());

        if let Some(depends) = depends_raw {
            let deps: Vec<String> = depends
                .split(',')
                .map(|d| d.trim().to_string())
                .filter(|d| !d.is_empty())
                .collect();
            for dep in deps {
                if !steps.contains_key(&dep) {
                    warnings.push(format!("Step \"{id}\" depends on unknown step \"{dep}\""));
                    continue;
                }
                if let Some(step) = steps.get_mut(id) {
                    step.depends_on.push(dep.clone());
                }
                if let Some(step_dep) = steps.get_mut(&dep) {
                    step_dep.depended_on_by.push(id.clone());
                }
            }
        }
    }

    let mut in_degree: HashMap<String, usize> = HashMap::new();
    for (id, step) in &steps {
        in_degree.insert(id.clone(), step.depends_on.len());
    }

    let mut execution_order: Vec<Vec<String>> = Vec::new();
    let mut processed: HashSet<String> = HashSet::new();
    let total = steps.len();
    let mut guard = 0usize;

    while processed.len() < total && guard < total + 1 {
        guard += 1;
        let mut batch: Vec<String> = steps
            .keys()
            .filter(|id| !processed.contains(*id) && in_degree.get(*id).copied().unwrap_or(0) == 0)
            .cloned()
            .collect();
        batch.sort();

        if batch.is_empty() {
            let mut remaining: Vec<String> = steps
                .keys()
                .filter(|id| !processed.contains(*id))
                .cloned()
                .collect();
            remaining.sort();
            warnings.push(format!(
                "Cycle detected involving: {}",
                remaining.join(", ")
            ));
            execution_order.push(remaining.clone());
            for id in remaining {
                processed.insert(id);
            }
            break;
        }

        execution_order.push(batch.clone());
        for id in batch {
            processed.insert(id.clone());
            if let Some(step) = steps.get(&id) {
                for dependent in &step.depended_on_by {
                    if let Some(v) = in_degree.get_mut(dependent) {
                        *v = v.saturating_sub(1);
                    }
                }
            }
        }
    }

    let mut entry_points: Vec<String> = steps
        .iter()
        .filter_map(|(id, step)| {
            if step.depends_on.is_empty() {
                Some(id.clone())
            } else {
                None
            }
        })
        .collect();
    entry_points.sort();

    let mut gate_positions = Vec::new();
    for (idx, batch) in execution_order.iter().enumerate() {
        if batch
            .iter()
            .any(|id| steps.get(id).map(|s| s.is_gate).unwrap_or(false))
        {
            gate_positions.push(idx);
        }
    }

    let has_terminal = steps.values().any(|s| s.is_terminal);

    WorkflowGraph {
        entry_points,
        steps,
        execution_order,
        gate_positions,
        has_terminal,
        warnings,
    }
}

fn collect_workflow_blocks(
    blocks: &[IntentBlock],
    inside_parallel: bool,
    out: &mut Vec<(IntentBlock, bool)>,
) {
    for block in blocks {
        let in_parallel = inside_parallel || block.block_type == "parallel";
        if WORKFLOW_TYPES.contains(&block.block_type.as_str()) {
            out.push((block.clone(), in_parallel));
        }
        if let Some(children) = &block.children {
            collect_workflow_blocks(children, in_parallel, out);
        }
    }
}
